core-backend/common/src/mock.rs

```diff
@@ -264,3 +267,94 @@ impl BackendExt for MockExt {
         Ok(())
     }
 }
+
+#[derive(Debug)]
+pub struct MockMemory {
+    pages: Vec<u8>,
+}
+
+impl MockMemory {
+    pub fn new(initial_pages: u32) -> Self {
+        let mut pages = Vec::with_capacity(initial_pages as usize * WASM_PAGE_SIZE);
+        unsafe {
+            pages.set_len(initial_pages as usize * WASM_PAGE_SIZE);
+        }
+        Self { pages }
+    }
+
+    fn page_index(&self, offset: u32) -> Option<usize> {
+        let index = offset as usize / WASM_PAGE_SIZE;
+        if (index as usize) < self.pages.len() / WASM_PAGE_SIZE {
+            Some(index as usize)
+        } else {
+            None
+        }
+    }
+}
+
+impl Memory for MockMemory {
+    type GrowError = ();
+
+    fn grow(&mut self, pages: WasmPage) -> Result<(), Self::GrowError> {
+        let new_size = self.pages.len() + (pages.raw() as usize) * WASM_PAGE_SIZE;
+        self.pages.reserve(new_size);
+        unsafe {
+            self.pages.set_len(new_size);
+        }
+        Ok(())
+    }
+
+    fn size(&self) -> WasmPage {
+        WasmPage::new((self.pages.len() / WASM_PAGE_SIZE) as u32).unwrap_or(WasmPage::default())
+    }
+
+    fn write(&mut self, offset: u32, buffer: &[u8]) -> Result<(), MemoryError> {
+        let page_index = self
+            .page_index(offset)
+            .ok_or(MemoryError::RuntimeAllocOutOfBounds)?;
+        let page_offset = offset as usize % WASM_PAGE_SIZE;
+
+        if page_offset as usize + buffer.len() > WASM_PAGE_SIZE {
+            return Err(MemoryError::RuntimeAllocOutOfBounds);
+        }
+
+        let page_start = page_index * WASM_PAGE_SIZE;
+        let start = page_start + page_offset as usize;
+
+        if start + buffer.len() > self.pages.len() {
+            return Err(MemoryError::RuntimeAllocOutOfBounds);
+        }
+
+        let dest = &mut self.pages[start..(start + buffer.len())];
+        dest.copy_from_slice(buffer);
+
+        Ok(())
+    }
+
+    fn read(&self, offset: u32, buffer: &mut [u8]) -> Result<(), MemoryError> {
+        let page_index = self
+            .page_index(offset)
+            .ok_or(MemoryError::AccessOutOfBounds)?;
+        let page_offset = offset as usize % WASM_PAGE_SIZE;
+
+        if page_offset as usize + buffer.len() > WASM_PAGE_SIZE {
```

Comment:
Isn't `page_offset` already `usize` here?
```suggestion
        if page_offset + buffer.len() > WASM_PAGE_SIZE {
```

Reply:
Done